README.md
# Erroran

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/iamprincesly/erroran/Release?style=flat-square) ![NPM](https://img.shields.io/npm/l/erroran?style=flat-square) ![npm](https://img.shields.io/npm/v/erroran?style=flat-square) ![npm](https://img.shields.io/npm/dm/erroran?style=flat-square) ![GitHub Repo stars](https://img.shields.io/github/stars/iamprincesly/erroran?style=flat-square) [![Rate on Openbase](https://badges.openbase.com/js/rating/erroran.svg)](https://openbase.com/js/erroran?utm_source=embedded&utm_medium=badge&utm_campaign=rate-badge) [![kandi X-Ray](https://kandi.openweaver.com/badges/xray.svg)](https://kandi.openweaver.com/javascript/iamprincesly/erroran) ![Twitter Follow](https://img.shields.io/twitter/follow/iamprincesly?style=flat-square)

A simple error handling middleware made for [Express](http://expressjs.com/) app. It can be used for throwing operational error in JSON response to the client. Erroran also catch all programming errors in production and throw friendly/desired error message to client.

-   [Installation](#installation)
-   [Usage](#usage)
    -   [Simple Usage](#simple-usage)
    -   [More Options](#more-options)
-   [Configuration](#configuration)
-   [Programming Errors](#programming-errors)
-   [Coming Soon](#coming-soon)
-   [License](#license)
-   [Author](#author)

## Installation

This is a [Node.js](https://nodejs.org/en/) module available through the
[npm registry](https://www.npmjs.com/). Installation is done using the
[`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```sh
$ npm install erroran
```

## Usage

### Simple Usage

Using the Erroran class to return operational error

```javascript
var express = require('express');
var { Erroran, ErroranHandler } = require('erroran');
var app = express();

app.get('/products/:id', function (req, res, next) {
    return next(
        new Erroran('You are not logged in! Please log in to get access', 401)
    );
});

// Make sure this is always your last middleware
app.use(ErroranHandler());

app.listen(5000, () => {
    console.log(`Server running in ${process.env.NODE_ENV} mode on port 5000`);
});
```

Alternatively, you can throw operational error like this

```javascript
var express = require('express');
var { Erroran, ErroranHandler } = require('erroran');
var app = express();

app.get('/products/:id', function (req, res, next) {
    throw new Erroran(
        'You are not logged in! Please log in to get access',
        401
    );
});

// Make sure this is always your last middleware
app.use(ErroranHandler());

app.listen(5000, () => {
    console.log(`Server running in ${process.env.NODE_ENV} mode on port 5000`);
});
```
Sample response
```json
{
    "status": "failed",
    "message": "You are not logged in! Please log in to get access"
}
```

NOTE: If you are throwing Erroran in async/await function make sure you catch the error object in ```catch()``` block and pass it to the ```next()``` function like below.

```javascript
var express = require('express');
var { Erroran, ErroranHandler } = require('erroran');
var app = express();

app.get('/users/:id', async function (req, res, next) {
    try {
        var user = await User.findById(req.params.id);

        if (!user) {
            throw new Erroran('There is no user with that id', 404);
        }
    } catch (err) {
        next(err);
    }
});

// Make sure this is always your last middleware
app.use(ErroranHandler());

app.listen(5000, () => {
    console.log(`Server running in ${process.env.NODE_ENV} mode on port 5000`);
});
```

If you throw Erroran in a `then()` chain you can as well catch the error in `catch()` block and pass it to the `next()` function like below.

```javascript
User.findById(req.params.id)
    .then((user) => {
        if (user.role !== 'admin') {
            throw new Erroran('You don\'t have permission to access this route', 403);
        }
    })
    .catch((err) => {
        next(err);
    });
```

### More Options

Erroran gives more options for throwing operational errors without instantiating the Erroran class. You can throw any of the Erroran static methods below with or without ```string``` as an argument.

```javascript
var { Erroran } = require('erroran');

Erroran.badRequest(); // 400 status code
Erroran.notAuthenticated(); // 401 status code
Erroran.paymentRequired(); // 402 status code
Erroran.forbidden(); // 403 status code
Erroran.notFound(); // 404 status code
Erroran.unsupportedHTTPMethod(); // 405 status code
Erroran.notAcceptable(); // 406 status code
Erroran.proxyAuthenticationRequired(); // 407 status code
Erroran.requestTimeout(); // 408 status code
Erroran.conflict(); // 409 status code
Erroran.gone(); // 410 status code
Erroran.lengthRequired(); // 411 status code
Erroran.preconditionFailed(); // 412 status code
Erroran.requestEntityTooLarge(); // 413 status code
Erroran.requestURITooLong(); // 414 status code
Erroran.unsupportedMediaType(); // 415 status code
Erroran.requestedRangeNotSatisfiable(); // 416 status code
Erroran.expectationFailed(); // 417 status code;
Erroran.internalServerError(); // 500 status code
Erroran.notImplemented(); // 501 status code
Erroran.badGateway(); // 502 status code
Erroran.serviceUnavailable(); // 503 status code
Erroran.gatewayTimout(); // 504 status code
Erroran.unsupportedHTTPVersion(); // 505 status code
```
You can pass your desire error message to any of the Erroran static methods as an argument, example below.

```javascript
var { Erroran } = require('erroran');

throw Erroran.notAuthenticated('Please login to access this resource');
```
Alternatively, you can return it to the ```next() ```function

```javascript
var express = require('express');
var { Erroran } = require('erroran');

var app = express();

app.get('/products/:id', function (req, res, next) {
    return next(Erroran.notAuthenticated('Please login to access this resource'));
});
```

## Configuration

Erroran has support for MongoDB, JWT and programming error. It catches MongoDB `'CastError'`, `'ValidationError'` and duplicate key `'11000'` error and also catches JWT `'JsonWebTokenError'` and `'TokenExpiredError'` in production.

You can configure Erroran by passing an object as an argument to the middleware like below.

```javascript
var express = require('express');
var { Erroran, ErroranHandler } = require('erroran');
var app = express();

var prodErrMsg = {
    ProgrammingErrorMsg: 'My programming errors in production', // When other programming error occur in production
    CastErrorMsg: 'My MongoDB cast error message in production', // MongoDB CastError
    MongoDBDuplicateKeyErrorMsg: 'My MongoDB duplicate key error message in production', // MongoDB DuplicateKeyError 11000
    ValidationErrorMsg: 'My validation error message in production', // MongoDB ValidationError
    JsonWebTokenErrorMsg: 'My Json Web token error message in production', // JWT JsonWebTokenError
    JsonTokenExpiredErrorMsg: 'My Json Web token Expiring error message in production', // TokenExpiredError
};

// Make sure this is always your last middleware
app.use(ErroranHandler(prodErrMsg));

app.listen(5000, () => {
    console.log(`Server running in ${process.env.NODE_ENV} mode on port 5000`);
});
```

Erroran has default messages to throw whenever MongoDB, JWT or programming error occur in production.

### Programming errors

For any programming error, Erroran will send full error stack and error message in development with 500 status code while sending `'Something went wrong, please try again later'` or whatever error message you define in `'ProgrammingErrorMsg'` with 500 status code in production.

All errors are sent in JSON response.

## Coming soon

-   Erroran will be able to send error and data to view
-   Erroran will be able to log all programming errors that occur in production into a log file

## License

[MIT License](http://www.opensource.org/licenses/mit-license.php)

## Author

[Sylvanus Etim](https://github.com/iamprincesly) ([iamprincesly@gmail.com](mailto:iamprincesly@gmail.com))


HISTORY.md
2.0.1 / 2023-08-03
==================
  * Fix bug

2.0.0 / 2022-1-25
==================
  * Changed ```getwayTimeout``` method to ```gatewayTimeout```

1.0.4 / 2022-1-8
==================
  * Added support for ```'EvalError'```, ```'RangeError'```, ```'URIError'```, ```'AggregateError'``` and ```'InternalError'``` erorr name

1.0.3 / 2022-1-3
==================
  * Added support for ```'Error'``` erorr name

1.0.2 / 2021-12-28
==================
  * Added support for ```'ReferenceError'``` and ```'TypeError'```

1.0.1 / 2021-12-27
==================
  * Fixed ```'SyntaxError'``` error in production

1.0.0 / 2021-12-26
==================
  * Catches all programming errors
  * Added support for MongoDB and JWT
  * Added more HTTP error status codes
  * Update documentation

0.1.3 / 2021-12-16
==================
  * Fixed programming error
  * Update documentation

0.1.1 / 2021-12-03
==================
  * Fixed status code for throwing `forbidden` error

0.1.0 / 2021-11-25
==================
  * Initial release

test/Erroran.test.js
/**
 * Erroran
 * @author Sylvanus Etim <iamprincesly@gmail.com>
 * @copyright Copyright (c) 2021 Sylvanus Etim
 * @license MIT Licensed
 */
var { Erroran } = require('../index');

describe('Erroran', () => {
    // Test for bad request
    it('Should throw bad request error', () => {
        expect(() => {
            Erroran.badRequest('This is an error from bad request');
        }).toThrow();
    });

    // Test for unauthenticated
    it('Should throw unauthentication error', () => {
        expect(() => {
            Erroran.notAuthenticated('This is an unauthentication erorr');
        }).toThrow();
    });

    // Test for payment required
    it('Should throw payment required error', () => {
        expect(() => {
            Erroran.paymentRequired('This is payment required erorr');
        }).toThrow();
    });

    // Test for forbidden error
    it('Should throw forbiden error', () => {
        expect(() => {
            Erroran.forbidden('This is forbidden erorr');
        }).toThrow();
    });

    // Test for not found error
    it('Should throw not found error', () => {
        expect(() => {
            Erroran.notFound('This is not found erorr');
        }).toThrow();
    });

    // Test for unsupported HTTP method error
    it('Should throw unsupported HTTP method error', () => {
        expect(() => {
            Erroran.unsupportedHTTPMethod(
                'This is unsupported HTTP method erorr'
            );
        }).toThrow();
    });

    // Test for not acceptable error
    it('Should throw not acceptable error', () => {
        expect(() => {
            Erroran.notAcceptable('This is not acceptable erorr');
        }).toThrow();
    });

    // Test for proxy authentication required error
    it('Should throw proxy authentication required error', () => {
        expect(() => {
            Erroran.proxyAuthenticationRequired(
                'This is proxy authentication required erorr'
            );
        }).toThrow();
    });

    // Test for request timeout error
    it('Should throw request timeout error', () => {
        expect(() => {
            Erroran.requestTimeout('This is request timeout erorr');
        }).toThrow();
    });

    // Test for conflict error
    it('Should throw conflict error', () => {
        expect(() => {
            Erroran.conflict('This is conflict erorr');
        }).toThrow();
    });

    // Test for gone error
    it('Should throw gone error', () => {
        expect(() => {
            Erroran.gone('This is gone erorr');
        }).toThrow();
    });

    // Test for length required error
    it('Should throw length required error', () => {
        expect(() => {
            Erroran.lengthRequired('This is length required erorr');
        }).toThrow();
    });

    // Test for precondition failed error
    it('Should throw precondition failed error', () => {
        expect(() => {
            Erroran.preconditionFailed('This is precondition failed erorr');
        }).toThrow();
    });

    // Test for request entity too large error
    it('Should throw request entity too large error', () => {
        expect(() => {
            Erroran.requestEntityTooLarge(
                'This is request entity too large erorr'
            );
        }).toThrow();
    });

    // Test for request URI too long error
    it('Should throw request URI too long error', () => {
        expect(() => {
            Erroran.requestURITooLong('This is request URI too long erorr');
        }).toThrow();
    });

    // Test for unsupported media type error
    it('Should throw unsupported media type error', () => {
        expect(() => {
            Erroran.unsupportedMediaType(
                'This is unsupported media type erorr'
            );
        }).toThrow();
    });

    // Test for requested range not satisfiable error
    it('Should throw requested range not satisfiable error', () => {
        expect(() => {
            Erroran.requestedRangeNotSatisfiable(
                'This is requested range not satisfiable erorr'
            );
        }).toThrow();
    });

    // Test for expectation failed error
    it('Should throw expectation failed error', () => {
        expect(() => {
            Erroran.expectationFailed('This is expectation failed erorr');
        }).toThrow();
    });

    // Test for internal server error
    it('Should throw internal server error', () => {
        expect(() => {
            Erroran.internalServerError('This is internal server erorr');
        }).toThrow();
    });

    // Test for not implemented error
    it('Should throw not implemented error', () => {
        expect(() => {
            Erroran.notImplemented('This is not implemented erorr');
        }).toThrow();
    });

    // Test for bad gateway error
    it('Should throw bad gateway error', () => {
        expect(() => {
            Erroran.badGateway('This is bad gateway erorr');
        }).toThrow();
    });

    // Test for service unavailable error
    it('Should throw service unavailable error', () => {
        expect(() => {
            Erroran.serviceUnavailable('This is service unavailable erorr');
        }).toThrow();
    });

    // Test for gateway timeout error
    it('Should throw gateway timeout error', () => {
        expect(() => {
            Erroran.gatewayTimout('This is gateway timeout erorr');
        }).toThrow();
    });

    // Test for unsupported HTTP version error
    it('Should throw unsupported HTTP version error', () => {
        expect(() => {
            Erroran.unsupportedHTTPVersion(
                'This is unsupported HTTP version erorr'
            );
        }).toThrow();
    });
});


demo.js
/**
 * Erroran
 * @author Sylvanus Etim <iamprincesly@gmail.com>
 * @copyright Copyright (c) 2021 Sylvanus Etim
 * @license MIT Licensed
 */

'use strict';

var express = require('express');
var { Erroran, ErroranHandler } = require('erroran');
var app = express();

// Returning Erroran to next() function
app.get('/returning/products/:id', function (req, res, next) {
   return next(
       new Erroran('You are not logged in! Please log in to get access', 401)
   );
});

// Throwing Erroran
app.get('/throwing/products/:id', function (req, res, next) {
    throw new Erroran(
        'You are not logged in! Please log in to get access',
        401
    );
});

// In async/await function
app.get('/users/:id', async function (req, res, next) {
   try {
       var user = await User.findById(req.params.id);

       if (!user) {
           throw new Erroran('There is no user with that id', 404);
       }
   } catch (err) {
       next(err);
   }
});

// In then() chaining 
app.get('/then/users/:id', function (req, res, next) {
   User.findById(req.params.id)
    .then((user) => {
        if (user.role !== 'admin') {
            throw new Erroran('You don\'t have permission to access this route', 403);
        }
    })
    .catch((err) => {
        next(err);
    });
});

/**
 * You can use any of the below Erroran static methods
 * to throw an error to save the stress of passing status code as an argument
 */
Erroran.badRequest(); // 400 status code
Erroran.notAuthenticated(); // 401 status code
Erroran.paymentRequired(); // 402 status code
Erroran.forbidden(); // 403 status code
Erroran.notFound(); // 404 status code
Erroran.unsupportedHTTPMethod(); // 405 status code
Erroran.notAcceptable(); // 406 status code
Erroran.proxyAuthenticationRequired(); // 407 status code
Erroran.requestTimeout(); // 408 status code
Erroran.conflict(); // 409 status code
Erroran.gone(); // 410 status code
Erroran.lengthRequired(); // 411 status code
Erroran.preconditionFailed(); // 412 status code
Erroran.requestEntityTooLarge(); // 413 status code
Erroran.requestURITooLong(); // 414 status code
Erroran.unsupportedMediaType(); // 415 status code
Erroran.requestedRangeNotSatisfiable(); // 416 status code
Erroran.expectationFailed(); // 417 status code;
Erroran.internalServerError(); // 500 status code
Erroran.notImplemented(); // 501 status code
Erroran.badGateway(); // 502 status code
Erroran.serviceUnavailable(); // 503 status code
Erroran.gatewayTimout(); // 504 status code
Erroran.unsupportedHTTPVersion(); // 505 status code

/**
 * Can pass your desire error message as an argument
 * any of the above Erroran static methods like below.
 */
throw Erroran.notAuthenticated('Please login to access this resource');

// Alternatively, you return it to the next function as well
app.get('/returning/products/:id', function (req, res, next) {
   return next(Erroran.notAuthenticated('Please login to access this resource'));
});

/**
 * You can config it like below
 */
var prodErrMsg = {
   ProgrammingErrorMsg: 'My programming errors in production', // When other programming error occur in production
   CastErrorMsg: 'My MongoDB cast error message in production', // MongoDB CastError
   MongoDBDuplicateKeyErrorMsg: 'My MongoDB duplicate key error message in production', // MongoDB DuplicateKeyError 11000
   ValidationErrorMsg: 'My validation error message in production', // MongoDB ValidationError
   JsonWebTokenErrorMsg: 'My Json Web token error message in production', // JWT JsonWebTokenError
   JsonTokenExpiredErrorMsg: 'My Json Web token Expiring error message in production', // TokenExpiredError
};

// Make sure this is always your last middleware
app.use(ErroranHandler(prodErrMsg)); // with configuration
app.use(ErroranHandler()); // without configuration

app.listen(5000, () => {
    console.log(`Server running in ${process.env.NODE_ENV} mode on port 5000`);
});


index.js
/*
 Copyright (c) 2021 Sylvanus Etim <iamprincesly@gmail.com>

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
 notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
 notice, this list of conditions and the following disclaimer in the
 documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
 derived from this software without specific prior written permission.
 */
'use strict';

/**
 * Export Erroran main class and erroran error
 * handling middleware
 */
module.exports = {
    Erroran: require('./libs/erroran'),
    ErroranHandler: require('./libs/handler'),
};

libs/handler.js
/**
 * Erroran
 * @author Sylvanus Etim <iamprincesly@gmail.com>
 * @copyright Copyright (c) 2021 Sylvanus Etim
 * @license MIT Licensed
 */

'use strict';

var helpers = require('./helpers');

/**
 * Export error middleware.
 */
module.exports = ErroranHandler;

/**
 * This is the main Erroran error handling middleware.
 * Pass it to express app as the last middleware
 *
 * @param {*} [options]
 * @return {function} (err, req, res, next)
 */
function ErroranHandler(options) {
    var opts = options || {};

    var appEnv = process.env.NODE_ENV || 'development';

    // Set default programming error message in production
    var proErrorMsg = opts.ProgrammingErrorMsg || 'Something went wrong, please try again';

    return function errorMiddleware(err, req, res, next) {
        if (appEnv === 'production') {
            var error = { ...err };

            var errorMessage;

            error.proErrorMsg = proErrorMsg;
            error.message = err.message;

            /**
             * Throw programming error in production.
             * Catches 'Error', 'SyntaxError', 'RangeError', 'URIError',
             * 'AggregateError', 'InternalError', 'EvalError', 'TypeError',
             * 'ReferenceError', and 'ErroranInvalidArgument' error names at the moment.
             * Looking forward to add more in the future as I discovered them.
             */
            var proErorrNames = [
                'Error',
                'SyntaxError',
                'RangeError',
                'URIError',
                'AggregateError',
                'InternalError',
                'EvalError',
                'TypeError',
                'ReferenceError',
                'ErroranInvalidArgument',
            ];

            if (proErorrNames.includes(err.name)) {
                error = helpers.programmingError(proErrorMsg);
            }

            /**
             * Throw CastErrorMsg in production
             */
            if (err.name === 'CastError') {
                errorMessage = opts.CastErrorMsg;
                error = helpers.CastError(err, errorMessage);
            }

            /**
             * Throw MongoDBDuplicateKeyErrorMsg in production
             */
            if (err.code === 11000) {
                errorMessage = opts.MongoDBDuplicateKeyErrorMsg;
                error = helpers.MongoDBDuplicateFieldsError(err, errorMessage);
            }

            /**
             * Throw ValidationErrorMsg in production
             */
            if (err.name === 'ValidationError') {
                errorMessage = opts.ValidationErrorMsg;
                error = helpers.ValidationError(err, errorMessage);
            }

            /**
             * Throw JsonWebTokenErrorMsg in production
             */
            if (err.name === 'JsonWebTokenError') {
                errorMessage = opts.JsonWebTokenErrorMsg;
                error = helpers.JWTMalfunctionError(errorMessage);
            }

            /**
             * Throw JsonTokenExpiredErrorMsg in production
             */
            if (err.name === 'TokenExpiredError') {
                errorMessage = opts.JsonTokenExpiredErrorMsg;
                error = helpers.JWTExpiredError(errorMessage);
            }

            helpers.handleProdError(error, req, res);
        } else {
            helpers.handleDevError(err, req, res);
        }
    };
}


libs/Erroran.js
/**
 * Erroran
 * @author Sylvanus Etim <iamprincesly@gmail.com>
 * @copyright Copyright (c) 2021 Sylvanus Etim
 * @license MIT Licensed
 */

'use strict';

/**
 * Main class
 *
 * @class Erroran
 * @extends {Error}
 */
module.exports = class Erroran extends Error {
    static BAD_REQUEST = 400;
    static NOT_AUTHENTICATED = 401;
    static PAYMENT_REQUIRED = 402;
    static FORBIDDEN = 403;
    static NOT_FOUND = 404;
    static UNSUPPORTED_METHOD = 405;
    static NOT_ACCEPTABLE = 406;
    static PROXY_AUTHENTICATION_REQUIRED = 407;
    static REQUEST_TIMEOUT = 408;
    static CONFLICT = 409;
    static GONE = 410;
    static LENGTH_REQUIRED = 411;
    static PRECONDITION_FAILED = 412;
    static REQUEST_ENTITY_TOO_LARGE = 413;
    static REQUEST_URI_TOO_LONG = 414;
    static UNSUPPORTED_MEDIA_TYPE = 415;
    static REQUESTED_RANGE_NOT_SATISFIABLE = 416;
    static EXPECTATION_FAILED = 417;
    static INTERNAL_SERVER_ERROR = 500;
    static NOT_IMPLEMENTED = 501;
    static BAD_GATEWAY = 502;
    static SERVICE_UNAVAILABLE = 503;
    static GATEWAY_TIMEOUT = 504;
    static UNSUPPORTED_HTTP_VERSION = 505;

    constructor(message, statusCode = 500, name, operational = true) {
        super(message);
        this.name = name || this.constructor.name;
        this.statusCode = statusCode;
        this.status = `${statusCode}`.startsWith('4') ? 'failed' : 'error';
        this.isOperational = operational ? true : false;
    }

    static #_invalidArgument(method) {
        throw new this(
            'Invalid arugment passed to: ' + method + '() static method',
            500,
            'ErroranInvalidArgument', false
        );
    }

    /**
     * Automatically send '400' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Bad Request]
     * @throw {object} this
     * @memberof Erroran
     */
    static badRequest(message = 'Bad Request') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('badRequest');
        }

        throw new this(message, this.BAD_REQUEST);
    }

    /**
     * Automatically send '401' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Not authenticated']
     * @throw {object} this
     * @memberof Erroran
     */
    static notAuthenticated(message = 'Not authenticated') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('notAuthenticated');
        }

        throw new this(message, this.NOT_AUTHENTICATED);
    }

    /**
     * Automatically send '402' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Payment Required']
     * @throw {object} this
     * @memberof Erroran
     */
    static paymentRequired(message = 'Payment Required') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('paymentRequired');
        }

        throw new this(message, this.PAYMENT_REQUIRED);
    }

    /**
     * Automatically send '403' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Forbidden']
     * @throw {object} this
     * @memberof Erroran
     */
    static forbidden(message = 'Forbidden') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('forbidden');
        }

        throw new this(message, this.FORBIDDEN);
    }

    /**
     * Automatically send '404' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Not Found]
     * @throw {object} this
     * @memberof Erroran
     */
    static notFound(message = 'Not Found') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('notFound');
        }

        throw new this(message, this.NOT_FOUND);
    }

    /**
     * Automatically send '405' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='HTTP Method Not Allowed']
     * @throw {object} this
     * @memberof Erroran
     */
    static unsupportedHTTPMethod(message = 'HTTP Method Not Allowed') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('unsupportedHTTPMethod');
        }

        throw new this(message, this.UNSUPPORTED_METHOD);
    }

    /**
     * Automatically send '406' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Not Acceptable']
     * @throw {object} this
     * @memberof Erroran
     */
    static notAcceptable(message = 'Not Acceptable') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('notAcceptable');
        }

        throw new this(message, this.NOT_ACCEPTABLE);
    }

    /**
     * Automatically send '407' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Proxy Authentication Required']
     * @throw {object} this
     * @memberof Erroran
     */
    static proxyAuthenticationRequired(
        message = 'Proxy Authentication Required'
    ) {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('proxyAuthenticationRequired');
        }

        throw new this(message, this.PROXY_AUTHENTICATION_REQUIRED);
    }

    /**
     * Automatically send '408' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Request Timeout']
     * @throw {object} this
     * @memberof Erroran
     */
    static requestTimeout(message = 'Request Timeout') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('requestTimeout');
        }

        throw new this(message, this.REQUEST_TIMEOUT);
    }

    /**
     * Automatically send '409' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Conflict']
     * @throw {object} this
     * @memberof Erroran
     */
    static conflict(message = 'Conflict') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('conflict');
        }

        throw new this(message, this.CONFLICT);
    }

    /**
     * Automatically send '410' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Gone']
     * @throw {object} this
     * @memberof Erroran
     */
    static gone(message = 'Gone') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('gone');
        }

        throw new this(message, this.GONE);
    }

    /**
     * Automatically send '411' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Length Required']
     * @throw {object} this
     * @memberof Erroran
     */
    static lengthRequired(message = 'Length Required') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('lengthRequired');
        }

        throw new this(message, this.LENGTH_REQUIRED);
    }

    /**
     * Automatically send '412' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Precondition Failed']
     * @throw {object} this
     * @memberof Erroran
     */
    static preconditionFailed(message = 'Precondition Failed') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('preconditionFailed');
        }

        throw new this(message, this.PRECONDITION_FAILED);
    }

    /**
     * Automatically send '413' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Request Entity Too Large']
     * @throw {object} this
     * @memberof Erroran
     */
    static requestEntityTooLarge(message = 'Request Entity Too Large') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('requestEntityTooLarge');
        }

        throw new this(message, this.REQUEST_ENTITY_TOO_LARGE);
    }

    /**
     * Automatically send '414' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Request-URI Too Long']
     * @throw {object} this
     * @memberof Erroran
     */
    static requestURITooLong(message = 'Request-URI Too Long') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('requestURITooLong');
        }

        throw new this(message, this.REQUEST_URI_TOO_LONG);
    }

    /**
     * Automatically send '415' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Unsupported Media Type']
     * @throw {object} this
     * @memberof Erroran
     */
    static unsupportedMediaType(message = 'Unsupported Media Type') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('unsupportedMediaType');
        }

        throw new this(message, this.UNSUPPORTED_MEDIA_TYPE);
    }

    /**
     * Automatically send '416' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Requested Range Not Satisfiable']
     * @throw {object} this
     * @memberof Erroran
     */
    static requestedRangeNotSatisfiable(
        message = 'Requested Range Not Satisfiable'
    ) {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('requestedRangeNotSatisfiable');
        }

        throw new this(message, this.REQUESTED_RANGE_NOT_SATISFIABLE);
    }

    /**
     * Automatically send '417' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Expectation Failed']
     * @throw {object} this
     * @memberof Erroran
     */
    static expectationFailed(message = 'Expectation Failed') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('expectationFailed');
        }

        throw new this(message, this.EXPECTATION_FAILED);
    }

    /**
     * Automatically send '500' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Internal Server Error']
     * @throw {object} this
     * @memberof Erroran
     */
    static internalServerError(message = 'Internal Server Error') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('internalServerError');
        }

        throw new this(message, this.INTERNAL_SERVER_ERROR);
    }

    /**
     * Automatically send '501' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Not Implemented']
     * @throw {object} this
     * @memberof Erroran
     */
    static notImplemented(message = 'Not Implemented') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('notImplemented');
        }

        throw new this(message, this.NOT_IMPLEMENTED);
    }

    /**
     * Automatically send '502' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Bad Gateway']
     * @throw {object} this
     * @memberof Erroran
     */
    static badGateway(message = 'Bad Gateway') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('badGateway');
        }

        throw new this(message, this.BAD_GATEWAY);
    }

    /**
     * Automatically send '503' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Service Unavailable']
     * @throw {object} this
     * @memberof Erroran
     */
    static serviceUnavailable(message = 'Service Unavailable') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('serviceUnavailable');
        }

        throw new this(message, this.SERVICE_UNAVAILABLE);
    }

    /**
     * Automatically send '504' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='Gateway Timeout']
     * @throw {object} this
     * @memberof Erroran
     */
    static gatewayTimout(message = 'Gateway Timeout') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('gatewayTimout');
        }

        throw new this(message, this.GATEWAY_TIMEOUT);
    }

    /**
     * Automatically send '505' status code
     * and your desire message to the client
     * @static
     * @param {string} [message='HTTP Version Not Supported']
     * @throw {object} this
     * @memberof Erroran
     */
    static unsupportedHTTPVersion(message = 'HTTP Version Not Supported') {
        if (typeof message !== 'string') {
            return this.#_invalidArgument('unsupportedHTTPVersion');
        }

        throw new this(message, this.UNSUPPORTED_HTTP_VERSION);
    }
};
